File: javaPractice/src/algoproblem/ReverseLinkedList.java
package algoproblem;

/**
 * Created by bing1 on 2019/2/12.
 */

//Definition for singly-linked list.


public class ReverseLinkedList {

    public ListNode reverseList(ListNode head) {
        ListNode p =head;
        ListNode lastNode =null;

        while(p !=null){
            ListNode temNode=p.next;
            p.next = lastNode;
            lastNode = p;
            p = temNode;

        }
        return lastNode;

    }

    public static void main(String[] args) {
        ListNode node5 = new ListNode(5);

        ListNode node4 = new ListNode(4);
        node4.next=node5;
        ListNode node3 = new ListNode(3);


        node3.next = node4;

        System.out.println(node3);

    }
}

File: pyPractice/algoproblem/reverse_linkedlist.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


'''
Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
'''
class Solution:
    def reverseList(self, head: 'ListNode') -> 'ListNode':
        p = head
        last_node = None

        while(p):
            temp = p.next
            p.next = last_node
            last_node = p
            p = temp

        return last_node



a = ListNode(5)
b = ListNode(4)
c = ListNode(3)